commit 1f08e84ff642294e42d138442a388989ffb20865
Author: Igor Perminov <igor.perminov@inbox.ru>
Date:   Tue Sep 22 00:25:44 2009 +0400

    mac80211: Fix [re]association power saving issue on AP side
    
    Consider the following step-by step:
    1. A STA authenticates and associates with the AP and exchanges
    traffic.
    2. The STA reports to the AP that it is going to PS state.
    3. Some time later the STA device goes to the stand-by mode (not only
    its wi-fi card, but the device itself) and drops the association state
    without sending a disassociation frame.
    4. The STA device wakes up and begins authentication with an
    Auth frame as it hasn't been authenticated/associated previously.
    
    At the step 4 the AP "remembers" the STA and considers it is still in
    the PS state, so the AP buffers frames, which it has to send to the STA.
    But the STA isn't actually in the PS state and so it neither checks
    TIM bits nor reports to the AP that it isn't power saving.
    Because of that authentication/[re]association fails.
    
    To fix authentication/[re]association stage of this issue, Auth, Assoc
    Resp and Reassoc Resp frames are transmitted disregarding of STA's power
    saving state.
    
    N.B. This patch doesn't fix further data frame exchange after
    authentication/[re]association. A patch in hostapd is required to fix
    that.
    
    Signed-off-by: Igor Perminov <igor.perminov@inbox.ru>
    Signed-off-by: John W. Linville <linville@tuxdriver.com>

diff --git a/net/mac80211/tx.c b/net/mac80211/tx.c
index 5143d203256b..fd4028296613 100644
--- a/net/mac80211/tx.c
+++ b/net/mac80211/tx.c
@@ -367,7 +367,10 @@ ieee80211_tx_h_unicast_ps_buf(struct ieee80211_tx_data *tx)
 	struct ieee80211_hdr *hdr = (struct ieee80211_hdr *)tx->skb->data;
 	u32 staflags;
 
-	if (unlikely(!sta || ieee80211_is_probe_resp(hdr->frame_control)))
+	if (unlikely(!sta || ieee80211_is_probe_resp(hdr->frame_control)
+			|| ieee80211_is_auth(hdr->frame_control)
+			|| ieee80211_is_assoc_resp(hdr->frame_control)
+			|| ieee80211_is_reassoc_resp(hdr->frame_control)))
 		return TX_CONTINUE;
 
 	staflags = get_sta_flags(sta);